api/provenance/provenance.go

```diff
@@ -6,57 +6,79 @@ package provenance
 import (
 	"fmt"
 	"runtime"
+	"runtime/debug"
 	"strings"
 )
 
-var (
-	version = "unknown"
-	// sha1 from git, output of $(git rev-parse HEAD)
-	gitCommit = "$Format:%H$"
-	// build date in ISO8601 format, output of $(date -u +'%Y-%m-%dT%H:%M:%SZ')
-	buildDate = "1970-01-01T00:00:00Z"
-	goos      = runtime.GOOS
-	goarch    = runtime.GOARCH
-)
-
 // Provenance holds information about the build of an executable.
 type Provenance struct {
 	// Version of the kustomize binary.
-	Version string `json:"version,omitempty"`
+	Version string `json:"version,omitempty" yaml:"version,omitempty"`
 	// GitCommit is a git commit
-	GitCommit string `json:"gitCommit,omitempty"`
+	GitCommit string `json:"gitCommit,omitempty" yaml:"gitCommit,omitempty"`
+	// GitTreeState is the state of the git tree
+	GitTreeState string `json:"gitTreeState,omitempty" yaml:"gitTreeState,omitempty"`
 	// BuildDate is date of the build.
-	BuildDate string `json:"buildDate,omitempty"`
+	BuildDate string `json:"buildDate,omitempty" yaml:"buildDate,omitempty"`
 	// GoOs holds OS name.
-	GoOs string `json:"goOs,omitempty"`
+	GoOs string `json:"goOs,omitempty" yaml:"goOs,omitempty"`
 	// GoArch holds architecture name.
-	GoArch string `json:"goArch,omitempty"`
+	GoArch string `json:"goArch,omitempty" yaml:"goArch,omitempty"`
+	// GoVersion holds Go version.
+	GoVersion string `json:"goVersion,omitempty" yaml:"goVersion,omitempty"`
 }
 
 // GetProvenance returns an instance of Provenance.
 func GetProvenance() Provenance {
-	return Provenance{
-		version,
-		gitCommit,
-		buildDate,
-		goos,
-		goarch,
+	p := Provenance{
+		BuildDate:    "unknown",
+		Version:      "unknown",
+		GitCommit:    "unknown",
+		GitTreeState: "unknown",
+		GoOs:         runtime.GOOS,
+		GoArch:       runtime.GOARCH,
+		GoVersion:    runtime.Version(),
+	}
+	info, ok := debug.ReadBuildInfo()
+	if !ok {
+		return p
+	}
+
+	// override with values from BuildInfo
+	if info.Main.Version != "" {
+		p.Version = info.Main.Version // TODO: Figure out whether tagging alone gets this set correctly
+	}
+	p.GoVersion = info.GoVersion
+	for _, setting := range info.Settings {
+		switch setting.Key {
+		case "vcs.revision":
+			p.GitCommit = setting.Value
+		case "vcs.modified":
+			switch setting.Value {
+			case "true":
+				p.GitTreeState = "dirty" // TODO: This ends up being "dirty" in goreleaser for some reason.
+			case "false":
+				p.GitTreeState = "clean"
+			}
+		case "vcs.time": // TODO: This is the commit time, not the build time.
```

Comment:
🔥 this change as well. The build date used by kubectl really is the build date, not the commit date. I think it comes from here, ultimately: https://github.com/kubernetes/kubernetes/blob/f33ddd1b9c1f53ca9bb7dd8a55c7e15ced1292a8/hack/lib/version.sh#L168